### DISCUSSION TITLE
Bulk create/edit Shop Pack Item through CSV

### DISCUSSION THREAD

**Discussion by TDodgeCo** [gameserverapp/Platform#889] [Suggestions]
### **Overview**

This proposal suggests the implementation of a bulk Shop Pack Item CSV uploader within the GameServerApp.com dashboard. The feature aims to streamline the creation of multiple shop packs by allowing users to upload a CSV file with the necessary data, which the system can then parse and process to create shop packs in bulk.

### **Problem Statement**

Currently, creating shop pack items individually through the GameServerApp.com dashboard is time-consuming, especially when dealing with dozens or even hundreds of items. The existing process involves manually inputting details for each item, which can be inefficient and prone to errors.

### **Proposed Solution**

The proposed solution is to introduce a CSV uploader that allows users to define multiple shop pack items in a single file. This feature would enable users to:

1. Prepare the shop pack item data in a CSV format.
2. Upload the CSV file through the dashboard interface.
3. The system parses the CSV and creates the shop pack items with the specified details.

### **CSV Structure**

The CSV file should include columns for all necessary fields required to create a shop pack item. Example columns might include:

- **Item Name**: The name of the shop pack item.
- **Description**: A brief description of the item.
- **Label**: Custom label for the item.
- **Token Price**: The price in tokens for the item.
- **Visibility**: Whether the item is visible to all or restricted to specific roles.
- **Image URL** (optional): A URL to the item's image.
- **Additional Settings**: Any other configurable settings available in the dashboard.

### **Key Benefits**

1. **Efficiency**: Reduces the time and effort required to create multiple shop pack items.
2. **Accuracy**: Minimizes the risk of manual entry errors.
3. **User Convenience**: Especially useful for users managing large inventories of items, making the platform more user-friendly.

### **Implementation Considerations**

- **Data Validation**: The system should include robust validation to ensure the data in the CSV is correctly formatted and complete.
- **Error Handling**: Clear error messages and logs should be provided if the CSV contains errors, with instructions for users on how to correct them.
- **Security**: Ensure that the uploader and processing logic are secure to prevent malicious data uploads.

### **Conclusion**

Implementing a bulk Shop Pack Item CSV uploader would significantly enhance the user experience on the GameServerApp.com dashboard, particularly for users managing extensive inventories. It aligns with the goals of increasing efficiency and user satisfaction, making it a valuable addition to the platform.

**Comment by IIXx6IIxX** [comment]
Nicely presented. 10000% on board. Anything to not have to do these individually would be a huge improvement.

**Comment by Ian-Highlander** [comment]
Massively on board with this suggestion. Big big BIG thumbs up from me, if I could press that thumbsup button 1000 times I would lol.

I have hundreds of items across shop packs for both ASE and ASA and it was a slow, painful process adding them all. Not a complaint as the web interface is well designed and intuitive but the time cost alone is insane.

Someone's obviously a project manager, written up like a pro. Great suggestion 👌

**Comment by babsgosgens** [comment]
Thank you for suggesting this.

Indeed, very well presented. 

I would add - if a pack id is present in the csv, the pack item is updated rather than created.

Continuing this line of thought, if the csv were to be imported into a collection… I guess I should stop now lol. If this were to be considered for addition at some point, I do have more ideas.

**Comment by NobleKangaroo** [comment]
Awesome idea. I'd like to add that a URL to an image is already present on GSA if an image has been uploaded, and can be used. However, an alternative option could be to use base64 encoding on imports to avoid potentially making hundreds of HTTP/S requests, depending on how this is implemented. URLs are decent, but the potential to bog down the service with hundreds of HTTP/S requests to external URLs is a consideration.

**Reply by goosenation** [reply]
A valid point. The images need to be sourced from somewhere. There are some free image to base64 encoding services online that would allow you to do that so they can be inserted as strings. 

Alternatively, they would have to be fed through some image service, but most are either subscription based, or have limited capacity. Shop images are not that big, so perhaps they would suffice. 

I store my shop images online for lack of a better solution. Tinified, obviously. I would prefer to be able to upload WEPP images, but since GSA is already converting images to WEBP I understand why that is not an upload option.

**Comment by mvaessen** [comment]
Hi folks, 

Thanks for the suggestion and your feedback!

After spending some time thinking about this one, here is what is currently the plan (this may change, but probably wont):
- The CSV feature will start as an add-on only feature, meaning only dashboards with an active subscription can use this. 

- Most of the Shop fields will be available in the CSV.
- You will be able to edit the actions, via the CSV too. This will use GSA's delivery format. This will be a bit more advanced to work with, but will let you tap into the raw delivery code.
- Uploads (images) are most likely not going to be a thing. It is fairly complicated to determine whether the given image was already uploaded with the current setup, meaning GSA would have to redownload on each CSV import. This is likely becoming a huge data bandwidth drain, and therefor poses a risk to the stability, so I cant allow that.


As with most updates, we'll likely start with a simple first version, and then add more options on the second version.

**Reply by babsgosgens** [reply]
Sounds good Max. I'm looking forward to test driving it!

**Comment by mvaessen** [comment]
Hi folks,

This feature is now available on your dashboard.
Please reach out on discord if you have any questions or need help.

![Screenshot 2024-10-28 at 17 05 56](https://github.com/user-attachments/assets/1334963b-1c50-495e-ae02-d675c4e68b7d)


----

Some packages may not load into Excel properly when using a comma-separated export, due to the delivery content also using comma's. To mitigate this can you export using ; / semicolon as a separator / delimiter. Because semicolon files don't load well out of the box, can you use the steps below to load them properly:

Opening ; separator files, in Excel:
1. Open new Excel sheet
2. Go to Data tab
3. Click Get data
4. Choose Text/CSV
5. Select the CSV export
6. Select ; / Semicolon as Delimiter and then load the data.
